src/System.Linq/src/System/Linq/Iterator.cs

```diff
@@ -67,5 +67,50 @@ void IEnumerator.Reset()
                 throw Error.NotSupported();
             }
         }
+
+        internal abstract class ManualIterator<TSource> : IEnumerable<TSource>, IEnumerator<TSource>
+        {
+            private readonly int _threadId;
+            internal int _state;
+
+            protected ManualIterator()
+            {
+                _threadId = Environment.CurrentManagedThreadId;
+            }
+
+            public abstract TSource Current { get; }
```

Comment:
I wonder if there are any cases where we're wrapping an iterator directly and the virtualness gets bypassed and this undoes it. Probably not, though.